### ISSUE TITLE
NotSupportedException in Mono.Cecil when using debug symbols

### ISSUE THREAD

**Issue by CreateAndInject** [icsharpcode/ILSpy#872]
Analyze any item, expand "Used By", double click any result:
<pre>
System.NotSupportedException
   at Mono.Cecil.Cil.InstructionOffset.get_Offset()
   at Mono.Cecil.Cil.CodeReader.ReadScope(ScopeDebugInformation scope)
   at Mono.Cecil.Cil.CodeReader.ReadDebugInfo()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody()
   at Mono.Cecil.Cil.CodeReader.ReadMethodBody(MethodDefinition method)
   at Mono.Cecil.MethodDefinition.<>c.<get_Body>b__41_0(MethodDefinition method, MetadataReader reader)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TRet& variable, TItem item, Func`3 read)
   at Mono.Cecil.MethodDefinition.get_Body()
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DecompileBody(MethodDefinition methodDefinition, IMethod method, EntityDeclaration entityDecl, ITypeResolveContext decompilationContext)
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.DoDecompile(MethodDefinition methodDefinition, IMethod method, ITypeResolveContext decompilationContext)
   at ICSharpCode.Decompiler.CSharp.CSharpDecompiler.Decompile(IList`1 definitions)
   at ICSharpCode.ILSpy.CSharpLanguage.DecompileMethod(MethodDefinition method, ITextOutput output, DecompilationOptions options)
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.DecompileNodes(DecompilationContext context, ITextOutput textOutput)
   at ICSharpCode.ILSpy.TextView.DecompilerTextView.<>c__DisplayClass33_0.<DecompileAsync>b__0()
</pre>

**Comment by dgrunwald** [comment]
That looks like a bug in Mono.Cecil.

As a workaround, you can disable "Use variable names from debug symbols" and then reload the assembly list (to reload the assemblies with debug symbols disabled).

**Comment by CreateAndInject** [comment]
But why there's no problem in ILSpy2.x？

**Comment by dgrunwald** [comment]
ILSpy2.x uses an older version of Mono.Cecil, maybe this is a new bug? Pretty much all the debug symbol support got updated in the new Mono.Cecil.

**Comment by CreateAndInject** [comment]
But there's no problem if I don't use "Analyze", I can click any item from TreeView, and no problem.

**Comment by dgrunwald** [comment]
I just managed to reproduce this problem with the ICSharpCode.Decompiler.Tests.exe assembly. The whole assembly can be disassembled/decompiled without issues, so this is indeed specific to the analyzers.

The analyzers use cecil from multiple threads. (multi-threaded decompilation when decompiling a whole project does the same; but there the cecil accesses only make up a small portion of the runtime, so races are less likely)
Maybe some parts of the new debug symbol stuff are not thread-safe?
We used to maintain our own thread-safe cecil branch, but that eventually got merged into cecil master. Maybe the changes in cecil 0.10 introduced new threading issues? cc @jbevain
